### PR TITLE
Added action remove selection

### LINKED ISSUE #95: Add "remove cursor" action
When using multiple cursors this would allow the user to remove a cursor that is contained in the given target

Could call it "give" (opposite of "take")

### PR DESCRIPTION
**Pull request by AndreasArvidsson**
Closes #95

### CODE REVIEW

File: src/actions/RemoveSelection.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+  async run([targets]: [TypedSelection[]]): Promise<ActionReturnValue> {
+    await runOnTargetsForEachEditor(targets, async (editor, targets) => {
+      editor.selections = editor.selections.filter(
+        (selection) =>
+          !targets.find(
+            (target) =>
+              target.selection.selection.intersection(selection) &&
+              !target.selection.selection.start.isEqual(selection.end) &&
+              !target.selection.selection.end.isEqual(selection.start)
```
Comment (pokey) [inline src/actions/RemoveSelection.ts:31]: I believe you're checking that the selection has non empty overlap? Can't you just ask the intersection whether it is empty? If not, I'd add a comment to clarify what this is doing

**Review by pokey — APPROVED** [review]
Looks good! Left one minor comment. The tests look really good. I see you have an example where the target strictly contains the selection. In that vein, I'd add a test for the converse:

- [ ] Test where selection strictly contains target

File: src/actions/RemoveSelection.ts (hunk 1 above)
Comment (AndreasArvidsson) [reply src/actions/RemoveSelection.ts]: Well you could do that. . . if it wasn't in the middle of the night :p
I have updated it now :)

File: src/actions/Deselect.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+      // Remove selections with a non-empty intersection
+      const newSelections = editor.selections.filter(
+        (selection) =>
+          !targets.find((target) => {
+            const intersection =
+              target.selection.selection.intersection(selection);
+            return intersection && !intersection.isEmpty;
+          })
+      );
+      // The editor requires at least one selection. Keep "primary" selection active
```
Comment (pokey) [inline src/actions/Deselect.ts:31]: Shouldn't we just let this be an error?

File: src/actions/Deselect.ts [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+      editor.selections = newSelections.length
+        ? newSelections
+        : [new Selection(editor.selection.active, editor.selection.active)];
+    });
+
+    return {
+      thatMark: targets.map((target) => target.selection),
```
Comment (pokey) [inline src/actions/Deselect.ts:38]: Do you think the `that` mark would be more useful if it were the selections that got dropped? We could probably just leave it for now and see how it works in practice but just a thought

File: src/test/suite/fixtures/recorded/actions/giveDrum.yml [hunk 4]
```diff
@@ -0,0 +1,28 @@
+spokenForm: give drum
```
Comment (pokey) [inline src/test/suite/fixtures/recorded/actions/giveDrum.yml:1]: Nice thorough tests 👍

File: src/actions/Deselect.ts [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+  async run([targets]: [TypedSelection[]]): Promise<ActionReturnValue> {
+    await runOnTargetsForEachEditor(targets, async (editor, targets) => {
+      // Remove selections with a non-empty intersection
+      const newSelections = editor.selections.filter(
+        (selection) =>
+          !targets.find((target) => {
+            const intersection =
+              target.selection.selection.intersection(selection);
+            return intersection && !intersection.isEmpty;
```
Comment (pokey) [inline src/actions/Deselect.ts:28]: Ok I had one more thought: what happens if selection is empty, eg just a cursor? In that case there's no way to deselect it right? Maybe we should allow intersection to be empty if selection was empty to begin with?

File: src/actions/Deselect.ts [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+  constructor(private graph: Graph) {
+    this.run = this.run.bind(this);
+  }
+
+  async run([targets]: [TypedSelection[]]): Promise<ActionReturnValue> {
+    await runOnTargetsForEachEditor(targets, async (editor, targets) => {
+      // Remove selections with a non-empty intersection
+      const newSelections = editor.selections.filter(
+        (selection) =>
+          !targets.find((target) => {
```
Comment (pokey) [inline src/actions/Deselect.ts:25]: Total nit here sorry, but

```suggestion
          !targets.some((target) => {
```

**Review by pokey — COMMENTED** [review]
Ok nice tests thanks. Left a few more comments. I think I found a case we're missing

File: src/actions/Deselect.ts (hunk 2 above)
Comment (AndreasArvidsson) [reply src/actions/Deselect.ts:31]: We could if you prefer that. If you have a really large selection just getting that out of the way with that the select command might be nice. It basically turns into an escape in that case

File: src/actions/Deselect.ts (hunk 3 above)
Comment (AndreasArvidsson) [reply src/actions/Deselect.ts:38]: I think we can leave it like this and try it out. In most cases that is token you actually referenced. But I do agree that could potentially be useful. 

Maybe we can use the source mark?

File: src/actions/Deselect.ts (hunk 5 above)
Comment (AndreasArvidsson) [reply src/actions/Deselect.ts:28]: That is a thought
you could just do a range over it but I get what you mean

Comment (AndreasArvidsson) [reply src/actions/Deselect.ts]: Implemented and it works quite well. You can reference a token to remove empty selections on both sides of it or you can combine with before and after to just remove on a specific side.

File: src/actions/Deselect.ts (hunk 2 above)
Comment (pokey) [reply src/actions/Deselect.ts:31]: Oh that's kinda cool i guess.  Feels logically incorrect but useful enough to let it slide 😊

File: src/actions/Deselect.ts (hunk 3 above)
Comment (pokey) [reply src/actions/Deselect.ts:38]: Hmm not sure source mark is right here but interesting idea.  Agreed tho let's just see if we ever actually want a way to refer to the removed selections in practice

**Pull request merged**